FSharp testing support (using Fuchu)

spencerwi opened the pull request:
Adding support for completing kata in F#, using [Fuchu](https://github.com/mausch/Fuchu) (an hspec-alike for F#)

spencerwi commented:
Bitten by Travis timeouts again. :\

jhoffner commented on lib/runners/fsharp.js:
```diff
@@ -16,7 +18,28 @@ module.exports.run = function(opts, cb)
         {
-            throw 'Test framework is not supported'
+            var runFixtureUsingFuchu = [
+                'let codeWarsTestPrinter: Fuchu.Impl.TestPrinters = {',
```
FYI you can use ES6 string templates here if that makes things easier

jhoffner commented:
F# has been a big desire from the community. Awesome!

jhoffner commented:
One more thing before I merge. Any idea of how to get the `<DESCRIBE::>` block in there? It looks like you used all of the Printer options available, but it does say in the documentation that tests can be grouped with arbitrary nesting. 

Not a huge deal but it would be nice if we can support nesting. If you do figure out how to get nesting working, you will also have to make sure to output a `<COMPLETEDIN::>` tag at the end of each IT/DESCRIBE test, which the runner parser uses to determine when a block is ended.

Kavignon commented:
Does it mean the F# language was fully integrated or are we waiting for me
to close the issue I opened on the subject?

On Aug 30, 2016 19:28, "Spencer Williams" <notifications@github.com> wrote:

> Adding support for completing kata in F#, using Fuchu
> <https://github.com/mausch/Fuchu> (an hspec-alike for F#)
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Codewars/codewars-runner-cli/pull/234
> Commit Summary
>
>    - FSharp testing support (using Fuchu)
>
> File Changes
>
>    - *M* README.md
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-0>
>    (2)
>    - *A* examples/fsharp.yml
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-1>
>    (20)
>    - *A* frameworks/fsharp/Fuchu/Fuchu.1.0.3.nupkg
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-2>
>    (0)
>    - *A* frameworks/fsharp/Fuchu/Fuchu.nuspec
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-3>
>    (18)
>    - *A* frameworks/fsharp/Fuchu/[Content_Types].xml
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-4>
>    (1)
>    - *A* frameworks/fsharp/Fuchu/_rels/.rels
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-5>
>    (1)
>    - *A* frameworks/fsharp/Fuchu/lib/Fuchu.XML
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-6>
>    (705)
>    - *A* frameworks/fsharp/Fuchu/lib/Fuchu.dll
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-7>
>    (0)
>    - *A* frameworks/fsharp/Fuchu/lib/Fuchu.pdb
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-8>
>    (0)
>    - *A* frameworks/fsharp/Fuchu/license.html
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-9>
>    (57)
>    - *A* frameworks/fsharp/Fuchu/package/services/metadata/
>    core-properties/b6186cf4f3cf4f74a896269c66cafa5e.psmdcp
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-10>
>    (1)
>    - *M* lib/runners/fsharp.js
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-11>
>    (27)
>    - *M* test/runners/fsharp_spec.js
>    <https://github.com/Codewars/codewars-runner-cli/pull/234/files#diff-12>
>    (30)
>
> Patch Links:
>
>    - https://github.com/Codewars/codewars-runner-cli/pull/234.patch
>    - https://github.com/Codewars/codewars-runner-cli/pull/234.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Codewars/codewars-runner-cli/pull/234>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AGbZ-qiLqime43dsFlZqKRFUlKLbxTqLks5qlF1tgaJpZM4Jwu52>
> .
>

spencerwi commented:
I'll need to dig some more to understand if I can extract a "DESCRIBE"; It looks like the printer hooks only get called on the per-test level, but I may be doing things wrong. I definitely do get the information for COMPLETEDIN, so I can add that.

spencerwi commented:
Okay, it turns out that Fuchu (the testing lib) wants quite a few dlls so that it can check for various other test libs' exception types. Added those, and added `<COMPLETEDIN::>` support, as well as a unit test for the failing-test case

spencerwi commented:
Alright, I give up on _not_ having Travis time out.

jhoffner commented:
I gotta write a script that somehow checks files and determines which runners need to be built. 

Thanks for trying. It passed once before so I'll go ahead and merge this and test locally. 

Awesome work!

jhoffner commented:
Had to run out of town unexpectedly so I'm a bit behind. Merging now and I'll be getting both this and ocaml merged into CW this week.

spencerwi commented:
It happens! No worries. Thanks for the work you do.

jhoffner commented:
@spencerwi Just wanted to give you a quick status update. I intended to have released this by now so that we could get some testing in before announcing. However these updates are wrapped up with some other ones and I haven't had the time yet to get some quirks worked out and fully tested. I do expect to get this out next week so that we can work through any issues that pop up and then ideally announce it the week after.

jhoffner commented:
I've deployed everything to staging but it looks like both languages are having issues. Even locally I am getting timeouts running F# tests, and for OCaml I am getting `spawn utop ENOENT` when I try to run the examples through my local runner. The mocha tests do run successfully though. 

I'm pretty sure I had the dotnet image building on my machine before so I'm not sure why its timing out all of a sudden. I'll keep digging.

jhoffner commented:
I take that back, it took me a few more tries but I got it to build. It looks like it runs pretty slow though, takes about 5 seconds just to run a simple example. Is F# naturally slow to compile or do you think something else might be causing this?

jhoffner commented:
OK, I got it working on preview and its decently quick. IDK whats going on with all of the inconsistencies but its been a wacky day.

spencerwi commented:
Eep, sorry for all the headaches! With OCaml I know they're was some weirdness where Docker's `RUN` doesn't set up the environment nicely, so I had to prefix a bunch of stuff with a bash script that runs opam's env-variable-setup stuff first (it works somewhat similarly to rbenv in Ruby or virtualenv in Python). 

I could see the OCaml image being slow to build, since it compiles some stuff from source (that's how opam does things). Once built, though, it should be very speedy to execute. 

F# I could see being slower; it's using mono, an unofficial-but-now-kinda-sorta-official cross-platform implementation of .NET, and since the .NET CLR is a whole runtime system, it's not as fast as something like C or Crystal or OCaml. It should be faster than Ruby or Python though. 

I'm sorry for all the bumps in the road. Codewars is really cool, and I hope I can help y'all out instead of just increasing overhead.

jhoffner commented:
No problem at all, your help is **greatly** appreciated. You can play with F# on staging now. OCaml is still having that issue so I will hold up on pushing that to production until we get things sorted. If you don't mind, please test out F# though and see if things work well enough. 

Don't publish anything yet on preview though. production users will see it and not be able to train on it. 

https://preview.codewars.com

Kavignon commented:
Hi,

I'd like to add you to my professional network on LinkedIn.

- Kevin

Confirm that you know Kevin: https://www.linkedin.com/comm/start/accept-invitation?sharedKey=9BuleU8c&invitationId=6208290945499217920&trk=eml-guest-invite-cta&trkEmail=eml-invite_guest-null-2-null-null-0%7Eivzc61vc%7Eio








You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=0-ivzc61vc-io&t=lun&midToken=AQFmrFVhGEd36Q&ek=invite_guest&loid=AQEBuOVpctPtfwAAAVihH8sEaIyoFHNs7eQdYDVG-q_w8LLcL0pb7IxC8J0o40Qe5SkWL-__q_RAfdlg5Ot8Zfc2inMJlIEv6re20R7DDGLD9uj7COgjz2xkvDovkiqTSeEAJsDopI2yh0pjK_Nnb5DmQNskPetWjeaiAyeIp4gIyk-pK4YZKaVKKee-QwI_Uw3URQx9HnCNHq-GtySyciMpnw&eid=0-ivzc61vc-io


This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=0-ivzc61vc-io&a=customerServiceUrl&ek=invite_guest

© 2016 LinkedIn Corporation, 1000 West Maude Avenue, Sunnyvale, CA 94085. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

Kavignon commented:
Kevin Avignon would like to connect on LinkedIn. How would you like to respond?

Accept: https://www.linkedin.com/comm/start/accept-invitation?sharedKey=9BuleU8c&invitationId=6208290945499217920&trk=eml-first_guest_reminder_01-hero-121-accept_text&trkEmail=eml-first_guest_reminder_01-hero-121-accept_text-null-%7Egdfwbx%7Eiw6msi7b%7Edl

View Kevin Avignon&#39;s profile: https://www.linkedin.com/comm/start/accept-invitation?sharedKey=9BuleU8c&invitationId=6208290945499217920&trk=eml-first_guest_reminder_01-hero-3-profile_text&trkEmail=eml-first_guest_reminder_01-hero-3-profile_text-null-%7Egdfwbx%7Eiw6msi7b%7Edl





You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=-gdfwbx-iw6msi7b-dl&t=lun&midToken=AQFmrFVhGEd36Q&ek=first_guest_reminder_01&li=123&m=unsub&ts=HTML&eid=-gdfwbx-iw6msi7b-dl&loid=AQEmYT6Y-XIZiAAAAVi8OAChtyUbzDWokZ_opo2mmM_fMOzqUPKeRoR4QBVWcbpfJalLlzp6N4cZyymhfa2NWW78aE9tif3bgerS8mH5ia_X5ZeYmN1zL4u8M-nVDapgUe5U7Rp26uYfvZCKzOJioFPNNG-MRQioQqYtBSorJ-Sq1A6R1kSmzV9pu9gCvSTkoGFzb0wxqvGiVo7skhN8vUW6DQ2_fw

This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=-gdfwbx-iw6msi7b-dl&a=customerServiceUrl&ek=first_guest_reminder_01

&copy; 2016 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.
